cfmm.mligo

```diff
@@ -293,6 +303,82 @@ let cash_transfer (storage : storage) (from : address) (to_ : address) (cash_amo
 #endif
 #endif
 
+(* Isoutility and Difference Equations *)
+// The Isoutility Function in https://hackmd.io/MkPSYXDsTf-giBprcDrc3w  
+let isoutility (target, cash, token : nat * nat * nat) : nat = 
+    let x = cash in 
+    let y = token in 
+    let a = target in 
+    let b = (Bitwise.shift_right 2n 48n) in // target is implicitly divided by 2 ** 48
+    let a2 = a * a in
+    let b2 = b * b in 
+    let ax2 = a2 * x * x in 
+    let by2 = b2 * y * y in
+    abs (a * x * b * y * (ax2 + by2) / (2 * a2 * b2))
+
+// Returns the price dy/dx of the isoutility function, i.e. a map by multiplication ∆x => ∆y, at a given point (x,y)
+let price_cash_to_token (target : nat) (cash : nat) (token : nat) : nat = 
+    let (x,y) = (cash, token) in
+    let a = target in 
+    let b = (Bitwise.shift_right 2n 48n) in // target is implicitly divided by 2 ** 48
+    let ax2 = x * x * a * a in
+    let by2 = y * y * b * b in
+    let num = y * (3n * ax2 + by2) in
+    let denom = x * (ax2 + 3n * by2) in
+    num/denom
+
+// A function to transfer assets while maintaining a constant isoutility
+let rec newton_dx_to_dy (x, y, dx, dy_approx, target, rounds : nat * nat * nat * nat * nat * nat) : nat = 
+    let a = target in 
+    let b = (Bitwise.shift_right 2n 48n) in // target is implicitly divided by 2 ** 48
+    let xp = x + dx in
+    let yp = y - dy_approx in 
+    let ax2 = a * a * x * x in let by2 = b * b * y * y in 
+    let axp2 = a * a * xp * xp in let byp2 = b * b * yp * yp in
+    (* Newton descent formula *)
+    let num = x * y * (ax2 + by2) - xp * yp * (axp2 + byp2) in 
+    let denom = xp * (axp2 + 3 * byp2) in
+    let adjust = (-num) / denom in 
+    // if (abs adjust = 0n) (* marginal difference calculated is rounded down to 0mutez *)
+    if (rounds <= 0n) (* Newton converges in 4 rounds, so we bound computation there *)
+    then 
+        let dy = dy_approx + adjust in 
+        abs dy // abs to make it a nat
+    else 
+        let new_dy_approx = abs (dy_approx - adjust) in
+        let next_round = abs (rounds - 1n) in
+        newton_dx_to_dy (x,y,dx,new_dy_approx,target,next_round)
+    (*
+        if denom = 0, then either:
+        1. xp = 0 => x + dx = 0, which we don't allow, or
+        2. a*xp = 0 and b*yp = 0 => a = 0 and (b = 0 or yp = 0), which implies
+           that the price target is 0.
+     *)
+
+// A function that outputs dy (diff_token) given x, y, and dx
+let trade_dcash_for_dtoken (x : nat) (y : nat) (dx : nat) (target : nat) (rounds : nat) : nat = 
+    let current_price = price_cash_to_token target x y in
+    let dy_approx = current_price * dx in
```

Comment:
This is not what we discussed, dy_approx should start at 0 so as to remain an underestimate of the true dy